spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/FilterPushdown.scala

```diff
@@ -0,0 +1,158 @@
+package org.apache.spark.sql.cassandra
+
+import com.datastax.spark.connector.cql.TableDef
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.sources.Filter
+
+
+/**
+ * Calculate the pushdown predicates for a given table and predicates.
+ *
+ * Partition pruning predicates are also detected an applied.
+ *  1. Only push down no-partition key column predicates with =, >, <, >=, <= predicate
+ *  2. Only push down primary key column predicates with = or IN predicate.
+ *  3. If there are regular columns in the pushdown predicates, they should have
+ *     at least one EQ expression on an indexed column and no IN predicates.
+ *  4. All partition column predicates must be included in the predicates to be pushed down,
+ *     only the last part of the partition key can be an IN predicate. For each partition column,
+ *     only one predicate is allowed.
+ *  5. For cluster column predicates, only last predicate can be non-EQ predicate
+ *     including IN predicate, and preceding column predicates must be EQ predicates.
+ *     If there is only one cluster column predicate, the predicates could be any non-IN predicate.
+ *  6. There is no pushdown predicates if there is any OR condition or NOT IN condition.
+ *  7. We're not allowed to push down multiple predicates for the same column if any of them
+ *     is equality or IN predicate.
+ *
+ */
+object FilterPushdown {
```

Comment:
If I ranked by preference right now

2
3
1

My rational being we don't know how stable these api's will be anyway and I feel like take the least effort for us to be compatible now.  I think 3 is a little cleaner for the code base. 1 is a lot of work which makes me feel like we are just chasing Spark Apis. I don't really have confidence that the Spark project will not to make changes big enough to totally disrupt the api to the point where we have to rewrite everything again.